### ISSUE TITLE
New Drag-and-Drop Utilities

### ISSUE THREAD

**Issue by Alyar666** [DevExpress/DevExtreme#9338]
# The Problem

 The DevExtreme library does not provide utilities for drag and drop.

 # The Proposed Solution

 We plan to add new **Draggable** and **Sortable** components. They allow you to handle drag and drop operations, sort lists, and move items between components.

 The **Draggable** API:

 ```js
let draggableOptions = {
    handle: string, // A CSS selector for the element that acts as the drag handle
    group: string, // Two Draggable components should have the same group value to allow dragging between them
    data: object, // A container for any data that can be useful during drag and drop gestures
    dragTemplate: template, // A template for the element displayed under the cursor during dragging
    clone: boolean, // If true, users drag clones of actual elements
    dragDirection: string, // The direction in which users can drag elements
    autoScroll: boolean, // If set to true, the Draggable component is auto-scrolled when dragging
    scrollSensitivity: number, // The distance in pixels from the viewport's edge after which the viewport starts to scroll
    scrollSpeed: number, // The speed at which the Draggable component is scrolled once the pointer reaches the scrollSensitivity distance
    boundary: object, // A CSS selector that determines the element beyond which the element being dragged cannot be moved
    cursorOffset: object, // The offset of the element being dragged relative to the cursor
    container: object, // The container in which to place the element being dragged
    onDragStart: function, // A function that is executed when dragging starts
    onDragMove: function, // A function that is executed when the pointer is moving
    onDragEnd: function, // A function that is executed when dragging ends
};
```
 The **Sortable** API includes the entire **Draggable** API plus the following members:
 ```js
let sortableOptions = {
    allowReordering: boolean,// Allows a user to reorder sortable items
    allowDropInsideItem: boolean,// Allows a user to drop an item inside another item
    filter: string, // A CSS selector for draggable elements
    itemOrientation: string, // Notifies the widget of the items' orientation
    dropFeedbackMode: string // Specifies the appearance of list elements while dragging
    onDragChange: function, // A function that is executed when an item's position in the DOM changes
    onReorder: function, // A function that is called when the draggable items are reordered
    onAdd: function, // A function that is called when a new item is added
    onRemove: function // A function that is called when a draggable item is removed
};
```
 ## Drag and drop items between lists
 ```js
$("#list1").dxSortable({
    filter: ">.item",
    data: "list1",
    group: "shared"
});
 $("#list2").dxSortable({
    filter: ">.item",
    data: "list2",
    group: "shared"
});
```
 ## Scheduler - Drag and drop an appointment from an external element
 ```js
$("#element").dxDraggable({
    data: "appointment",
    group: "shared",
    data: {...}
});
 $("#scheduler").dxScheduler({
    appointmentDragging: {
        group: "shared",
        onAdd: function(e) {
            $("#scheduler").dxScheduler("instance").addAppointment(e.itemData);
        }
    }
});
```
 For more information about cross-component drag and drop, refer to the [Record/Node Reordering in DataGrid and TreeList and Cross-Component Drag and Drop](https://github.com/DevExpress/DevExtreme/issues/9339) discussion page.
 # We Need Your Feedback
 ## Take a Quick Poll
 [Do you find the Draggable and Sortable components useful?](https://docs.google.com/forms/d/e/1FAIpQLSdZNr3UTV9v5wUMGtM5AipOCLdMy81n7V8JzXZjQ03L2y4fWg/viewform)
 ## Live Sandbox
Kanban: [JQuery](https://codepen.io/vconst/pen/rNNNZqw), [Angular](https://codesandbox.io/s/kanban-devextreme-sortable-8um4m), [Vue](https://codesandbox.io/s/kanban-devextreme-sortable-ceqjk), [React](https://codesandbox.io/s/kanban-devextreme-sortable-tnlfq)
TreeView node dragging: [JQuery](https://codepen.io/vconst/pen/mdddQVR), [Angular](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-6qf5j), [React](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-ecdb7)
 # Get Notified of Updates
 Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by b1tzer0** [comment]
Would you be able to drag from a DevExtreme component to an External Component?

I had to write a Knockout Component that would take a list item and allow me to drag it to a custom Map Editor that we wrote

**Comment by Alyar666** [comment]
Hello @b1tzer0,

Yes, you can do it. However, it will be necessary to track item movement over the external component. It will look something like this in the jQuery approach:

```
$("#externalComponent").externalComponent(...);
 ..... 
 $("#list").dxList({
    .....,
    itemDragging: {
        enabled: true,
    },
    onDragMove: function(e) {
        let $currentElement = $(document.elementFromPoint(e.clientX, e.clientY));

        if($currentElement.closest("#externalComponent").length){
            ...
        }
    },
    onDragEnd: function(e) {
        let $currentElement = $(document.elementFromPoint(e.clientX, e.clientY));

        if($currentElement.closest("#externalComponent").length){
            ...
        }
    }
});
```
Could you describe your scenario in greater detail so we can take it into account in our further development? A video showing your behavior in action might be helpful. It would also be very helpful if you could share an example that meets your specific needs.

**Comment by vconst** [comment]
We have published 19.2.2-beta packages. I have added some live examples to the main content. You're welcome to try them and share your feedback here.

**Comment by markeaston63** [comment]
Hi, I found a bug - but works great otherwise ... 

This does not work:

<dx-data-grid
    #datagrid
    id="gridContainer"
    [height]="600"
    [dataSource]="assemblyList"
    [showBorders]="true">
    <dxo-row-dragging
        [allowReordering]="true"
        [showDragIcons]="true"
        [onDragStart]="onDragStart"
        [onReorder]="onReorder"
    ></dxo-row-dragging>
    <dxo-scrolling mode="virtual"></dxo-scrolling>
    <dxo-sorting mode="none"></dxo-sorting>
    <dxo-editing mode="batch" [allowUpdating]="true" [allowDeleting]="false" [useIcons]="true"></dxo-editing>
    <dxi-column dataField="assemblyNumber"  [allowEditing]="false"  width="80px" fixed="true" caption="Assembly No"></dxi-column>
    <dxi-column dataField="assemblyType"  [allowEditing]="false"  caption="Aseembly Type"></dxi-column>
</dx-data-grid>

And this does work:

<dx-data-grid
    #datagrid
    id="gridContainer"
    [height]="600"
    [dataSource]="assemblyList"
    [showBorders]="true">
    <dxo-row-dragging
        [allowReordering]="true"
        [showDragIcons]="true"
        [onDragStart]="onDragStart"
        [onReorder]="onReorder"
    ></dxo-row-dragging>
    <dxo-scrolling mode="virtual"></dxo-scrolling>
    <dxo-sorting mode="none"></dxo-sorting>
    <dxo-editing mode="batch" [allowUpdating]="true" [allowDeleting]="false" [useIcons]="true"></dxo-editing>
    <dxi-column dataField="assemblyNumber"  [allowEditing]="false"  width="80px" caption="Assembly No"></dxi-column>
    <dxi-column dataField="assemblyType"  [allowEditing]="false"  caption="Aseembly Type"></dxi-column>
</dx-data-grid>

It fails on fixed="true"​

**Comment by woodreamz** [comment]
Do you plan to add the node dragging on TreeView for React? Or do we have a way to implement it in React by ourself?

Thanks!

**Comment by vconst** [comment]
@markeaston63, We have already fixed the issue in 19.2.3.

**Comment by vconst** [comment]
@woodreamz, 
It's possible to use our new dxSortable component to add the node dragging capability to TreeView for React. See the following example - [React](https://codesandbox.io/s/plain-data-structure-devextreme-tree-view-ecdb7)

**Comment by woodreamz** [comment]
@vconst Thank you!

**Comment by markeaston63** [comment]
Hi,
Im using 
19.2.3.

I have a grid on two separate tab panels, both with drag and drop. For some inexplicable reason when I start drag and drop on the grid in the second tab panel the tab panel underneath changes back to the first tab panel for me to drop the item I am dragging. I have carefully looked at my two grids, in two separate angular components and they are identified and bound uniquely - so I cant see any reason for this behaviour. 

Can you replicate this problem? would be easy to replicate - Tab Panel each with a datagrid that has drag and drop.
Cheers
Mark

**Comment by Alyar666** [comment]
Hello @markeaston63, 

I have tried to reproduce it based on your description, but on my side, the approach works fine (example: [https://codesandbox.io/s/devextreme-data-grid-54psz](https://codesandbox.io/s/devextreme-data-grid-54psz)). If the problem persists, please provide us with an example in the context of a new issue.

**Comment by Alyar666** [comment]
These features are now available in the [v19.2](https://github.com/DevExpress/DevExtreme/releases/tag/19.2.3) release. I'm closing this thread. In the case of bugs or questions, feel free to create new GitHub [issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
